File: backend/internal/services/report.go
package services

import (
	"akatengu/internal/model/db/report"
	"akatengu/internal/repos/query"
	"context"
	"fmt"
)

type ReportService struct {
	repo query.ReportRepo
}

func NewReportService(repo query.ReportRepo) *ReportService {
	return &ReportService{repo: repo}
}

func (r *ReportService) GetBalanceSheet(ctx context.Context, reportDate string) (*report.BalanceSheet, error) {
	return r.repo.GetBalanceSheet(ctx, reportDate)
}

func (r *ReportService) GetIncomeStatement(ctx context.Context, startDate, endDate string) (*report.IncomeStatement, error) {
	return r.repo.GetIncomeStatement(ctx, startDate, endDate)
}

func (r *ReportService) GetCashFlowStatement(ctx context.Context, startDate, endDate string) (*report.CashFlowStatement, error) {
	return r.repo.GetCashFlowStatement(ctx, startDate, endDate)
}

func (r *ReportService) GetDirectCashFlowStatement(ctx context.Context, startDate, endDate string) (*report.DirectCashFlowStatement, error) {
	return r.repo.GetDirectCashFlowStatement(ctx, startDate, endDate)
}

func (r *ReportService) GetEquityStatement(ctx context.Context, startDate, endDate string) (*report.EquityStatement, error) {
	return r.repo.GetEquityStatement(ctx, startDate, endDate)
}

func (r *ReportService) Verify(ctx context.Context, date string) error {
	bs, err := r.GetBalanceSheet(ctx, date)
	if err != nil {
		return err
	}

	is, err := r.GetIncomeStatement(ctx, date[:4]+"-01-01", date)
	if err != nil {
		return err
	}

	// 資產 = 負債 + 淨資產 + 本期損益
	right := bs.TotalLiabilities.Add(bs.TotalEquity).Add(is.NetIncome)
	if !bs.TotalAssets.Sub(right).IsZero() {
		return fmt.Errorf(
			"balance sheet equation broken: assets=%s, liabilities+equity+net=%s",
			bs.TotalAssets.StringFixed(2), right.StringFixed(2),
		)
	}
	return nil
}

File: frontside/src/api/report.ts
import { apiFetch } from './http';
import type { BalanceSheet, IncomeStatement, CashFlowStatement, DirectCashFlowStatement, EquityStatement } from '../types/report';

export const getBalanceSheet = async (reportDate: string): Promise<BalanceSheet> => {
  const response = await apiFetch(`/api/report/balance_sheet?report_date=${reportDate}`);
  if (!response.ok) {
    const problem = await response.json();
    throw new Error(problem.detail ?? problem.title ?? '查詢資產負債表失敗');
  }
  return response.json();
};

export const getIncomeStatement = async (beginDate: string, endDate: string): Promise<IncomeStatement> => {
  const response = await apiFetch(`/api/report/income_statement?begin_date=${beginDate}&end_date=${endDate}`);
  if (!response.ok) {
    const problem = await response.json();
    throw new Error(problem.detail ?? problem.title ?? '查詢損益表失敗');
  }
  return response.json();
};

export const getEquityStatement = async (beginDate: string, endDate: string): Promise<EquityStatement> => {
  const response = await apiFetch(`/api/report/equity_statement?begin_date=${beginDate}&end_date=${endDate}`);
  if (!response.ok) {
    const problem = await response.json();
    throw new Error(problem.detail ?? problem.title ?? '查詢權益變動表失敗');
  }
  return response.json();
};

export const getCashFlowStatement = async (beginDate: string, endDate: string): Promise<CashFlowStatement> => {
  const response = await apiFetch(`/api/report/cash_flow_statement?begin_date=${beginDate}&end_date=${endDate}`);
  if (!response.ok) {
    const problem = await response.json();
    throw new Error(problem.detail ?? problem.title ?? '查詢現金流量表失敗');
  }
  return response.json();
};

export const getDirectCashFlowStatement = async (beginDate: string, endDate: string): Promise<DirectCashFlowStatement> => {
  const response = await apiFetch(`/api/report/cash_flow_statement_direct?begin_date=${beginDate}&end_date=${endDate}`);
  if (!response.ok) {
    const problem = await response.json();
    throw new Error(problem.detail ?? problem.title ?? '查詢現金流量表（直接法）失敗');
  }
  return response.json();
};
